native/src/modules/endpoint/reducers/pushOrRefreshCategory.js

```diff
@@ -0,0 +1,149 @@
+import ErrorCodes from '../../error/ErrorCodes'
+
+const getAllAvailableLanguages = (
+  category: CategoryModel, city: string, language: string, cityLanguages: Array<LanguageModel>
+) => {
+  if (category.isRoot()) {
+    return new Map(cityLanguages.map(model => [model.code, `/${city}/${model.code}`]))
+  }
+  const allAvailableLanguages = new Map(category.availableLanguages)
+  allAvailableLanguages.set(language, category.path)
+  return allAvailableLanguages
+}
+
+const extractResultModelsAndChildren = (root: CategoryModel, categoriesMap: CategoriesMapModel, depth: number): {|
+  resultModels: { [path: PathType]: CategoryModel },
+  resultChildren: { [path: PathType]: $ReadOnlyArray<PathType> }
+|} => {
+  /**
+   * Extracts models and children from the (updated) categories map.
+   */
+  const resultModels = {}
+  const resultChildren = {}
+
+  forEachTreeNode(root, (node: CategoryModel) => categoriesMap.getChildren(node), depth, (node, children) => {
+    resultModels[node.path] = node
+    if (children) {
+      resultChildren[node.path] = children.map(child => child.path)
+    }
+  })
+  return {
+    resultModels,
+    resultChildren
+  }
+}
+
+const pushOrRefreshCategory = (state: CityContentStateType,
+  action: PushCategoryActionType | RefreshCategoryActionType): CityContentStateType => {
+  const type = action.type
+  const { categoriesMap, path, depth, key, language, city, resourceCache, cityLanguages } = action.params
+
+  // If there is an error in the old resourceCache, we want to override it
+  const newResourceCache = state.resourceCache.status === 'ready'
+    ? { ...state.resourceCache.value, ...resourceCache }
+    : resourceCache
+
+  const root: ?CategoryModel = categoriesMap.findCategoryByPath(path)
+
+  if (!root) {
+    const route: CategoryRouteStateType = {
+      path: path,
+      depth: depth,
+      language,
+      city,
+      status: 'error',
+      message: `Could not find a category with path '${path}'.`,
+      code: ErrorCodes.PageNotFound
+    }
+
+    return {
+      ...state,
+      categoriesRouteMapping: {
+        ...state.categoriesRouteMapping,
+        [key]: route
+      },
+      resourceCache: {
+        status: 'ready',
+        value: newResourceCache
+      },
+      searchRoute: { categoriesMap }
+    }
+  }
+
+  // Check whether another page in the same city is loading, e.g. because it is being refreshed.
+  // This is important for displaying the loading spinner.
+  const otherPageLoading = Object.values(state.categoriesRouteMapping)
+    // $FlowFixMe Flow does not support Object.values
+    .filter(route => city === route.city && path !== route.path && language === route.language)
+    // $FlowFixMe Flow does not support Object.values
+    .some(route => route.status === 'loading')
+
+  const { resultModels, resultChildren } = extractResultModelsAndChildren(root, categoriesMap, depth)
+
+  // $FlowFixMe Flow can't evaluate the status as it is dynamic
```

Comment:
`type` can have two values: `loading` and `ready` (and some more). As they are hard coded in the StateType.js, flow considers them "string literals". But the type of a line like e.g. `const x = condition ? 'loading' : 'ready'` is `string` instead of `string literal`, so flow complains.

There is probably a solution to this, like:
```
if (a) {
x = { type: 'loading', ...}
} else {
x = { type: 'ready', ...} 
}
```
This is more verbose than the inline solution and I don't like having to write such "unclean" code just because of flow. But can do it anyways if you want...